File: server/src/services/claude.service.js
/**
 * LeadPulse — AI Service (Google Gemini via @google/genai)
 *
 * Key design decisions:
 *  - Tries a list of fallback models in order if one is busy/unavailable.
 *  - Retries up to 3 times with 2-second delay between attempts.
 *  - getJSON returns parsed object; streamResponse writes SSE to res.
 *  - getText returns plain string.
 *
 * Free key: https://aistudio.google.com/apikey
 */

'use strict';

const { GoogleGenAI } = require('@google/genai');

// Ordered list — first available/non-busy model wins
// Confirmed working on this API key (AQ. prefix = GCP project key)
const MODELS = [
  'models/gemini-2.5-flash',      // primary — best quality
  'models/gemini-2.5-flash-lite', // fallback — lighter, still works
];

const MAX_RETRIES = 3;
const RETRY_DELAY = 2000; // ms

let _ai = null;

function getAI() {
  if (_ai) return _ai;
  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey || apiKey === 'your-gemini-api-key-here') {
    throw new Error(
      'GEMINI_API_KEY not set. Get a FREE key at https://aistudio.google.com/apikey'
    );
  }
  _ai = new GoogleGenAI({ apiKey });
  return _ai;
}

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Try a Gemini call across multiple models with retries.
 * @param {Function} fn - async fn(ai, model) => result
 */
async function withFallback(fn) {
  const ai = getAI();
  let lastErr;

  for (const model of MODELS) {
    for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
      try {
        return await fn(ai, model);
      } catch (err) {
        const status = err.status || err.code || 0;
        const isRetryable = status === 503 || status === 429 || status === 500;
        lastErr = err;

        if (!isRetryable) break; // wrong model / auth — skip to next model
        if (attempt < MAX_RETRIES) {
          console.warn(`[Gemini] ${model} attempt ${attempt} failed (${status}), retrying in ${RETRY_DELAY}ms...`);
          await sleep(RETRY_DELAY);
        }
      }
    }
    console.warn(`[Gemini] Skipping ${model}, trying next fallback...`);
  }

  throw lastErr;
}

/**
 * Stream a Gemini response as SSE to the HTTP response object.
 */
async function streamResponse(prompt, systemPrompt = '', res) {
  res.setHeader('Content-Type', 'text/event-stream');
  res.setHeader('Cache-Control', 'no-cache');
  res.setHeader('Connection', 'keep-alive');
  res.setHeader('Access-Control-Allow-Origin', process.env.CLIENT_URL || '*');
  res.flushHeaders?.();

  try {
    await withFallback(async (ai, model) => {
      const response = await ai.models.generateContentStream({
        model,
        config: {
          systemInstruction: systemPrompt || 'You are a helpful CRM assistant.',
          maxOutputTokens: 1200,
          temperature: 0.7,
        },
        contents: [{ role: 'user', parts: [{ text: prompt }] }],
      });

      for await (const chunk of response) {
        if (chunk.text) {
          res.write(`data: ${JSON.stringify({ text: chunk.text })}\n\n`);
        }
      }
    });
  } catch (err) {
    console.error('[Gemini streamResponse error]', err.message);
    res.write(`data: ${JSON.stringify({ text: '\n[AI temporarily unavailable — please try again in a moment]' })}\n\n`);
  }

  res.write('data: [DONE]\n\n');
  res.end();
}

/**
 * Get a structured JSON response from Gemini.
 */
async function getJSON(prompt, systemPrompt = '') {
  return withFallback(async (ai, model) => {
    const response = await ai.models.generateContent({
      model,
      config: {
        systemInstruction:
          (systemPrompt || 'You are a helpful assistant.') +
          ' IMPORTANT: Reply with valid JSON only. No markdown fences, no explanation.',
        maxOutputTokens: 800,
        temperature: 0.2,
        responseMimeType: 'application/json',
      },
      contents: [{ role: 'user', parts: [{ text: prompt }] }],
    });

    let text = response.text.trim();
    if (text.startsWith('```')) {
      text = text.replace(/```(?:json)?/g, '').trim();
    }
    return JSON.parse(text);
  });
}

/**
 * Get a plain-text (non-streaming) response from Gemini.
 */
async function getText(prompt, systemPrompt = '', maxTokens = 400) {
  return withFallback(async (ai, model) => {
    const response = await ai.models.generateContent({
      model,
      config: {
        systemInstruction: systemPrompt || 'You are a helpful CRM assistant.',
        maxOutputTokens: maxTokens,
        temperature: 0.5,
      },
      contents: [{ role: 'user', parts: [{ text: prompt }] }],
    });
    return response.text;
  });
}

/**
 * Log an AI feature call to the database (fire-and-forget).
 */
async function logAICall(feature, userId, input, output, tokensUsed = 0, contactId = null, dealId = null) {
  try {
    const { prisma } = require('../lib/prisma');
    await prisma.aiLog.create({
      data: {
        feature,
        userId,
        input: JSON.stringify(input),
        output: output ? JSON.stringify(output) : null,
        tokensUsed: tokensUsed || 0,
        contactId: contactId || null,
        dealId: dealId || null,
      },
    });
  } catch (e) {
    console.warn('[AI Log] Failed to write:', e.message);
  }
}

module.exports = { streamResponse, getJSON, getText, logAICall };

File: ml-service/services/claude_service.py
"""
Gemini AI service — uses google-genai (new official SDK).
Model: gemini-2.0-flash — FREE tier (15 req/min, 1M tokens/day).
Get free key: https://aistudio.google.com/apikey
"""

import os
import json

_client = None


def get_client():
    global _client
    if _client is None:
        api_key = os.getenv("GEMINI_API_KEY", "").strip()
        if not api_key or api_key == "your-gemini-api-key-here":
            raise RuntimeError(
                "GEMINI_API_KEY not set. Get a FREE key at "
                "https://aistudio.google.com/apikey and add it to your .env file"
            )
        from google import genai
        _client = genai.Client(api_key=api_key)
    return _client


MODEL      = "models/gemini-2.5-flash"
MINI_MODEL = "models/gemini-2.5-flash"

_SYSTEM_PROMPT = """You are LeadPulse's AI CRM assistant. You answer questions about the user's CRM data.

RULES:
1. Answer ONLY from the provided CRM data context below.
2. If the answer is not clearly present, say: "I don't have that specific information in your CRM data."
3. Be concise and specific. Reference contact names, deal titles, and dollar values directly.
4. Never fabricate CRM data, names, or numbers.
5. Format responses using plain text. Use bullet points for lists.
6. Keep answers under 200 words unless the user explicitly asks for detail."""


def stream_rag_answer(question: str, context: str, history: list = None):
    """
    Stream a RAG-grounded answer from Gemini as text chunks (generator).
    """
    from google import genai
    from google.genai import types

    client = get_client()

    # Build conversation history
    contents = []
    for turn in (history or [])[-6:]:
        role    = turn.get("role", "")
        content = turn.get("content", "")
        if role == "user":
            contents.append(types.Content(role="user",  parts=[types.Part(text=content)]))
        elif role in ("assistant", "model"):
            contents.append(types.Content(role="model", parts=[types.Part(text=content)]))

    user_msg = f"CRM Data Context:\n{context}\n\n---\n\nQuestion: {question}"
    contents.append(types.Content(role="user", parts=[types.Part(text=user_msg)]))

    response = client.models.generate_content_stream(
        model=MODEL,
        contents=contents,
        config=types.GenerateContentConfig(
            system_instruction=_SYSTEM_PROMPT,
            max_output_tokens=800,
            temperature=0.5,
        ),
    )

    for chunk in response:
        if chunk.text:
            yield chunk.text


def get_json_response(prompt: str, system_extra: str = "") -> dict:
    """Get a structured JSON dict from Gemini."""
    from google import genai
    from google.genai import types

    client = get_client()

    system = (
        "You are a JSON extraction assistant. "
        + system_extra
        + " IMPORTANT: Reply with valid JSON only. No markdown fences, no explanation."
    )

    response = client.models.generate_content(
        model=MODEL,
        contents=prompt,
        config=types.GenerateContentConfig(
            system_instruction=system,
            max_output_tokens=800,
            temperature=0.2,
            response_mime_type="application/json",
        ),
    )

    text = response.text.strip()
    if text.startswith("```"):
        text = text.replace("```json", "").replace("```", "").strip()

    return json.loads(text)


def get_text_response(prompt: str, system: str = "", max_tokens: int = 400) -> str:
    """Get a plain-text response from Gemini (non-streaming)."""
    from google import genai
    from google.genai import types

    client = get_client()

    response = client.models.generate_content(
        model=MINI_MODEL,
        contents=prompt,
        config=types.GenerateContentConfig(
            system_instruction=system or "You are a helpful CRM assistant.",
            max_output_tokens=max_tokens,
            temperature=0.5,
        ),
    )
    return response.text
